File: src/rpc.ts
import { NeovimClient, VimValue } from '@chemzqm/neovim'

import { BufferId, PreviewPayload } from './app-contract'

/**
 * Calls into the editor. Every entry point lives in
 * `lua/markdown-preview/rpc.lua`, so this file is the whole server-to-editor
 * contract.
 */
const invoke = (nvim: NeovimClient, method: string, args: VimValue[]) =>
  nvim.lua(`return require("markdown-preview.rpc").${method}(...)`, args)

export const echoError = (nvim: NeovimClient, lines: string[]) =>
  invoke(nvim, 'error', [lines]).catch(() => {
    // The editor is the thing we report errors to, so there is nowhere left to go.
  })

export const openUrl = (nvim: NeovimClient, url: string) =>
  invoke(nvim, 'open_url', [url])

export const previewPayload = async (
  nvim: NeovimClient,
  bufnr: BufferId,
  includeContent: boolean
) => await invoke(nvim, 'payload', [Number(bufnr), includeContent]) as PreviewPayload

File: lua/markdown-preview/rpc.lua
--- The surface the preview server calls back into Neovim with, over
--- `nvim_exec_lua`. Kept in one module so the TypeScript side (`src/rpc.ts`)
--- has a single contract to hold on to.
---
--- The opposite direction, editor to server, lives in `server.lua`.

local M = {}

---@param lines string[]
function M.error(lines)
  require("markdown-preview.util").error(lines)
end

---@param url string
function M.open_url(url)
  require("markdown-preview.browser").open(url)
end

---@param bufnr integer
---@param include_content boolean
---@return table
function M.payload(bufnr, include_content)
  return require("markdown-preview.preview").payload(bufnr, include_content)
end

return M
